sound/soc/sof/intel/hda.c

```diff
@@ -147,9 +147,21 @@ static int sdw_free_stream(struct device *dev,
 	return hda_ctrl_dai_widget_free(w, SOF_DAI_CONFIG_FLAGS_NONE, &data);
 }
 
+static int sdw_trigger(struct snd_pcm_substream *substream, int cmd, struct snd_soc_dai *dai)
+{
+	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(dai, substream->stream);
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(swidget->scomp);
+
+	if (sdev->pdata->ipc_type != SOF_INTEL_IPC4)
+		return 0;
+	return ipc4_be_dai_trigger(substream, cmd, dai);
```

Comment:
it's not super elegant to have tests for IPC4. we could make the sdw_callback non-const or something where we only do the test on initialization.

Reply:
Or we do more surgery and move all the DAI stuff back in the sound/soc/sof side.